rocrtst: Reduce host memory limit to 70%

Sundance636 opened the pull request:
## Motivation
Reducing the upper bound for rocrtstFunc.Memory_Max_Mem to 70% from 90% to help reduce test execution time.

Sundance636 committed:
rocrtst: Reduce host memory limit to 70%

Reducing the upper bound for rocrtstFunc.Memory_Max_Mem to 70% from
90% to help reduce test execution time.

Signed-off-by: Sunday Clement <Sunday.Clement@amd.com>

shwetagkhatri commented:
Hi Sunday,
LGTM. But does this fix your issue - SWDEV-552886 ? Am just thinking, if this is enough to reduce the execution time on all ASICs ? And we wont see such issue again? Or we can add a input parameter to reduce memory limit? To be able to do memory stress test during local testing (non CI based) .
Any thoughts, @dayatsin-amd ?

Sundance636 commented:
> Hi Sunday, LGTM. But does this fix your issue - SWDEV-552886 ? Am just thinking, if this is enough to reduce the execution time on all ASICs ? And we wont see such issue again? Or we can add a input parameter to reduce memory limit? To be able to do memory stress test during local testing (non CI based) . Any thoughts, @dayatsin-amd ?

Hi Shweta,
I do agree I am a little skeptical too, just because this same issue kinda just keeps sneaking back, there was a pretty much duplicate issue from about a year ago. And plus David's previous patch of lowering the max pool size from 95% to 90%, I had similar ideas of manually inputting a parameter to control the memory limit. as per David, very fragmented memory (on the CI machines) seems to be a big factor in causing these large allocations in taking a very long time.

cfreeamd commented:
LGTM
Thoughts:
- Maybe there could be a rocrtst flag that allows longer-running tests (could be used for this tests and other tests), but defaults to something that finishes relatively quickly. 
- Pragmatically, if 70% finishes in a reasonable amount of time, most of the time, it seems like it would be ok.
- The thing we could do (longer term) is add some rocrtst infrastructure to put such things in a thread with a timer, so that we can treat (too) long running tests separately, and not necessarily view it as a "hang". Then if we were to get time-outs, we could run these tests without time-limited threads and see if there is an actual problem, or if the test just takes a long time.

dayatsin-amd commented:
I thought I posted a comment here a few days ago but it seems I forgot to hit the publish button.

@shwetagkhatri raises a good point about users being to test closer to the actual pool limit in non-CI environments.

There is an internal patch that has not been sync'ed with this branch yet:
https://github.com/AMD-ROCm-Internal/ROCR-Runtime/commit/24b34842ff9f24551cd92d5e805cbfcb419bc2be

And we can use ROCRTST_LIMIT_POOL_SIZE env variable when the user wants to increase the limits. So I am thinking we cherry-pick that patch into this branch, and then:

If ROCRTST_LIMIT_POOL_SIZE is not specified, set pool limit at 70%.
If ROCRTST_LIMIT_POOL_SIZE is specified, set pool limit to 90% of ROCRTST_LIMIT_POOL_SIZE

Sundance636 committed:
rocrtst: Add ROCRTST_LIMIT_POOL_SIZE env var

Add environment variable to override the memory pool sizes when running
tests.

Co-authored-by: David Yat Sin <David.YatSin@amd.com>

dayatsin-amd commented on projects/rocr-runtime/rocrtst/common/common.cc:
```diff
@@ -405,6 +405,16 @@ hsa_status_t AcquirePoolInfo(hsa_amd_memory_pool_t pool,
+      std::cout << "Warning: Pool size override > than reported size (override:"
+        << pool_size_limit << " reported:" << pool_i->size << ")" << std::endl;
+    }
+    pool_i->size = pool_size_limit;
```
You will need pool_size_limit to be declared as a global variable and initialized to 0 (probably in common.h).

And then in the other section something similar to this:

float pool_size_limit_ratio = 1.0;
if (ag_type == HSA_DEVICE_TYPE_CPU) {
   pool_size_limit_ratio = pool_size_limit ? 0.9 : 0.7
}

uint64_t upper_bound = pool_size_limit_ratio * pool_sz;
